vscode/vscode/vscode.mli

```diff
@@ -792,82 +699,71 @@ module StatusBarItem : sig
   val dispose : t -> unit
 
   val disposable : t -> Disposable.t
-
-  val t_of_js : Ojs.t -> t
-
-  val t_to_js : t -> Ojs.t
 end
 
 module WorkspaceFoldersChangeEvent : sig
-  type t
+  include Js.T
 
   val added : t -> WorkspaceFolder.t list
 
   val removed : t -> WorkspaceFolder.t list
 
   val create :
     added:WorkspaceFolder.t list -> removed:WorkspaceFolder.t list -> t
-
-  val t_of_js : Ojs.t -> t
-
-  val t_to_js : t -> Ojs.t
 end
 
 module FormattingOptions : sig
-  type t
+  include Js.T
 
   val tabSize : t -> int
 
   val insertSpaces : t -> bool
 
   val create : tabSize:int -> insertSpaces:bool -> t
-
-  val t_of_js : Ojs.t -> t
-
-  val t_to_js : t -> Ojs.t
 end
 
 module Event : sig
-  type 'a t = listener:('a -> unit) -> Disposable.t
-
-  val t_of_js : (Ojs.t -> 'a) -> Ojs.t -> 'a t
+  type 'a t =
+       listener:('a -> unit)
+    -> ?thisArgs:Js.Any.t
+    -> ?disposables:Disposable.t list
+    -> unit
+    -> Disposable.t
 
-  val t_to_js : ('a -> Ojs.t) -> 'a t -> Ojs.t
+  module Make (T : Js.T) : Js.T with type t = T.t t
 end
 
 module EventEmitter : sig
-  type t
+  type 'a t
```

Comment:
It's a bit of a shame that we're losing polymorphism here. Now we can't have a single `EventEmitter.fire` that will work for all event emitters. We could mend by just putting the dictionary inside the event emitter.

```ocaml
type 'a t
val make : (module Js.T with type t = 'a)-> 'a t
```

It's a little more code in the binding, but at least people don't need to instantiate modules to use the API. This is how various data structures in base work. What do you think?

Reply:
I feel that passing around the `to_js`/`of_js` values inside of `'a t` would add significant overhead, both to the runtime and code necessary to manipulate such values.

The benefit of functors is that once they're instantiated, they work like bindings to normal types. Inside the functor body, bindings can be written with gen_js_api that automatically use `T.t_to_js`/`T.t_of_js`.

Using a custom representation that bundles those functions would require each method of the generic type to do the conversions, which would get verbose very quickly.

Reply:
It's true that it's a little slower, but I doubt that it will make much difference for us. I've used this method gainfully in native ocaml, and I'm pretty certain js optimizes such code even better.

The point about boilerplate is more important. It's true that it adds more unpacking, but this overhead will become meaningless compared to the overhead of multiple modules for every type.

For now we only have a few callers. Let's leave things as is